wgsl/index.bs

```diff
@@ -942,9 +1019,336 @@ a result value, such as the non-result-value effects of its subexpressions.
+* There is no overload of `log2` with an [=integer scalar=] formal parameter, so the [=Materialization Rule=] cannot be applied directly.
+* Instead, apply the [=Promotion Rule=] to convert the operand to the [=AbstractFloat=] value 32.0.
+* Now the [=Materialization Rule=] applies, converting the [=AbstractFloat=] value 32.0 to the [=f32=] value 32.0f.
+* The resulting computation is equivalent to `log2(32.0f)`.
+
+Example:  The expression `1 + 2.5` is analyzed as follows:
+* `1 + 2.5` is parsed as an addition operation with subexpressions [=AbstractInt=] value 1, and [=AbstractFloat=] value 2.5.
+* There is no overload for |e|+|f| where |e| is integral and |f| is floating point.
+* However, once we apply the [=Promotion Rule=] to convert the [=AbstractInt=] operand 1 to [=AbstractFloat=] 1.0, there
+     is exactly one overload that can apply, where both operands are [=f32=].
+* Now the [=Materialization Rule=] is applied, producing the an addition over the [=f32=] values for 1.0 and 2.5.
+* The resulting computation is equivalent to `1.0f + 2.5f`.
+
+Example:  `let x = 1 + 2.5;`
+* As above, we evaluate the initializer to be the expression `1.0f + 2.5f` with type [=f32=].
+* The effect of the declaration is as if it were written `let x: f32 = 1.0f + 2.5f;`
+
+Example:  `1u + 2.5` results in a [=shader-creation error=]:
+* The `1u` term is an expression of type [=u32=].
+* The `2.5` term is an expression of type [=AbstractFloat=].
+* There is no valid typing for the expression:
+    * There is no implicit conversion from a GPU-materialized integral type to a floating point type.
+    * No type rule matches *e*`+`*f* with *e* in an integral type, and *f* in a floating point type.
+
+### Constexpr Expressions ### {#constexpr-expr}
+
+Constxpr expressions support two operations:
+[[#parenthesized-expressions|parenthesization]], and unary negation (see [[#arithmetic-expr]]).
+
+Example:  `(42)` is analyzed as follows:
+* The term `42` is the [=AbstractInt=] value 42.
+* Surrounding that term with parentheses produces a new expression `(42)` that is
+    of type [=AbstractInt=] with value 42.
+
+Example:  `-5` is analyzed as follows:
+* The term `5` is the [=AbstractInt=] value 5.
+* Preceding that term with '`-`' produces a new expression `-5` that is
+    of type [=AbstractInt=] with value -5.
+
+Example:  `-2147483648` is analyzed as follows:
+* The term `2147483648` is the [=AbstractInt=] value 2147483648.
+    Note that this value does **not** fit in a 32-bit signed integer.
+* Preceding that term with '`-`' produces a new expression `-2147483648` that is
+    of type [=AbstractInt=] with value -2147483648.
+
+Example:  `let minint = -2147483648;` is analyzed as follows:
+* As above, `-2147483648` evaluates to a [=AbstractInt=] value -2147483648.
+* A [=let declaration=] requires the initializer to be [=constructible=].
+    So we must apply the [=Materialization Rule=].
```

Comment:
In "Materialization and Promotion", you say "delete this whole section". This is a different section. Is it correct to assume that the materialization rule is gone, superseded by overload resolution?